### PR TITLE
feat: Enable SuccessfulCommandTest by leveraging Testcontainers

### PR DESCRIPTION
**Pull request by HofmeisterAn**
Relates #4193.

## Changes

I am submitting this pull request to enable `SuccessfulCommandTest` by leveraging Testcontainers. The changes in this PR spin up instances of Microsoft SQL Server that the test can run against. By doing so, we can achieve a broader coverage, reliable and reproducible test results.

With the use of Testcontainers, we can be sure that the test environment is always in the same state, regardless of the underlying host system. This is achieved by running the test against a container, which is created on-demand and destroyed after the test has completed. In addition, it allows the test to run in isolation, reducing the risk of interference from other processes or applications.

To run the tests with Testcontainers, a Moby compatible engine such as Docker is required.

### CODE REVIEW

**Comment by cijothomas** [comment]
@HofmeisterAn Thanks. Could you sign the CLA agreement, which is a pre-req before we can consider any contributions. The link from the CLA bot comment has instructions. Your employer might need to be consulted for any company specific instructions.

**Comment by HofmeisterAn** [comment]
I was able to integrate Testcontainers, enable and run the tests within just a few minutes. However, I was not aware of the current pipeline configuration, which adds unnecessary overhead and requires additional configuration when using it with Testcontainers. I'm not sure how we should proceed with the tests. I think the best approach is to move away from the Docker Compose setup and simply run the tests.

Regarding the issue with strong-named assemblies, I am aware of it. The issue has been [addressed](https://github.com/dotnet/Docker.DotNet/pull/567), and I am waiting for a release of Docker.DotNet to resolve it. Although it may not be critical for tests, it's good to know that it has been taken care of.

**Comment by HofmeisterAn** [comment]
@cijothomas any feedback how to continue?

**Comment by cijothomas** [comment]
> @cijothomas any feedback how to continue?

Lets wait for the signing fix, so we can make the CI green before proceeding further.
I like the overall approach, and yes, we can get rid of existing manual docker spin ups once we fully move to this.

**Comment by HofmeisterAn** [comment]
> System.ArgumentException : Cannot detect the Docker endpoint. Use either the environment variables or the ~/.testcontainers.properties file to customize your configuration:

This issue relates to the Docker Compose environment. The container created by Compose does not include Docker itself. We can either adjust the [Compose configuration](https://dotnet.testcontainers.org/examples/compose/) (Docker Desktop requires the env variable too) by mounting the daemon socket, or we can remove Compose since it is not needed. In that case, you can simply run the tests.

File: .github/workflows/integration.yml
Comment (alanwest) [inline .github/workflows/integration.yml:24]: We can probably get rid of the separate workflow step now as these tests should just run with the rest of the unit tests.

File: test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientTests.cs
Comment (alanwest) [inline test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientTests.cs:45]: If `mcr.microsoft.com/azure-sql-edge` works on Apple silicon, we could use an expression like the following to decide which image to use

```csharp
RuntimeInformation.IsOSPlatform(OSPlatform.OSX) 
&& RuntimeInformation.ProcessArchitecture == Architecture.Arm64
```

**Review by alanwest — COMMENTED** [review]
@HofmeisterAn I like this! I've been curious about Testcontainers but hadn't had a chance to try them out yet.

File: .github/workflows/integration.yml
Comment (HofmeisterAn) [reply .github/workflows/integration.yml:24]: But you still need the filter. I have tried running all tests without the filter, but the tests never complete. It seems like the remaining tests are waiting for the Docker Compose environment to be available. Please note that the filter already exists and is being used [here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/d924663dc3d6bfdf737bc49ceaa1addcec90a2d6/test/OpenTelemetry.Instrumentation.SqlClient.Tests/docker-compose.yml#L20), and was not introduced by me. I believe that there is another stage where the remaining tests run (see the following [Dockerfile](https://github.com/open-telemetry/opentelemetry-dotnet/blob/d924663dc3d6bfdf737bc49ceaa1addcec90a2d6/test/OpenTelemetry.Instrumentation.SqlClient.Tests/Dockerfile)). Thanks to Testcontainers, all these configurations are not necessary anymore.

Please note that the initial goal was to enable the disabled SQL tests and not to refactor all tests, etc. in the first place.

File: test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientTests.cs
Comment (HofmeisterAn) [reply test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientTests.cs:45]: Right now, it requires a bit more abstraction. The modules do not share a common database base class or interface, but this is not much of a problem. Why not use Azure SQL Edge for all platforms?

File: .github/workflows/integration.yml
Comment (HofmeisterAn) [reply .github/workflows/integration.yml:24]: Oh, my bad. I believe you are correct in that the tests are already run with `build-test`. However, they fail on Windows because the Docker image is not available (compatible) for the Windows engine.

Comment (alanwest) [reply .github/workflows/integration.yml:24]: > However, they fail on Windows because the Docker image is not available (compatible) for the Windows engine.

I still haven't had a chance to look too closely at TestContainers for .NET. Is it that they just aren't compatible with Windows at all? Or is it specifically the SQL image that is not compatible?

Comment (alanwest) [reply .github/workflows/integration.yml:24]: The original integration tests only ran on linux anyways. So maybe what we could do is introduce a new xUnit attribute that will skip the test if not linux. Something similar to the `SkipUnlessEnvVarFoundTheory` but instead `SkipIfNotLinux`.

Comment (HofmeisterAn) [reply .github/workflows/integration.yml:24]: > I still haven't had a chance to look too closely at TestContainers for .NET. Is it that they just aren't compatible with Windows at all? Or is it specifically the SQL image that is not compatible?

Testcontainers for .NET supports both. I ran it primarily with WCOW to test legacy Windows applications and MSIs. This particular case is all about the Docker image. Multi arch images that follow the [Windows container version compatibility](https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility) run fine.

> The original integration tests only ran on linux anyways. So maybe what we could do is introduce a new xUnit attribute that will skip the test if not linux.

👍

File: test/OpenTelemetry.Tests/Shared/SkipOnWindowsRunnerTheoryAttribute.cs
Comment (alanwest) [inline test/OpenTelemetry.Tests/Shared/SkipOnWindowsRunnerTheoryAttribute.cs:31]: For tests with this attribute, what will happen when I run them locally?

Comment (HofmeisterAn) [reply test/OpenTelemetry.Tests/Shared/SkipOnWindowsRunnerTheoryAttribute.cs:31]: Hm, it looks like `RUNNER_OS` does not return what the [docs](https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables) says, I will check.

Comment (alanwest) [reply test/OpenTelemetry.Tests/Shared/SkipOnWindowsRunnerTheoryAttribute.cs:31]: Ideally, the attribute causes the test to be skipped in all scenarios where either the environment is not compatible (windows, mac) or docker is not running.

Comment (HofmeisterAn) [reply test/OpenTelemetry.Tests/Shared/SkipOnWindowsRunnerTheoryAttribute.cs:31]: The only two cases are Docker is not running or Windows engine is enabled (very rare, macOS is supported too). If you like I can add something like we are doing in Testcontainers. We check if the Docker CLI response with the correct [platform](https://github.com/testcontainers/testcontainers-dotnet/blob/710987a2562f7bb5a58d64bb74cba17cc8681db4/tests/Testcontainers.Commons/DockerCli.cs#L44-L48). This will only enable the tests for environments that are capable of. WDYT?

Comment (HofmeisterAn) [reply test/OpenTelemetry.Tests/Shared/SkipOnWindowsRunnerTheoryAttribute.cs]: Lets see if the recent changes work better. BTW, I was trying a few times to test the changes in advance (on CI too), but I have serious issues building the forked version. I run into [this](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3412) issue all the time.

**Comment by cijothomas** [comment]
@HofmeisterAn Consider making a small PR into this repo and get it merged, so that you'll no longer considered "New contriutor". (New contributors require one of the maintainers to trigger the CI builds adding delays to you ability to test quikly)

File: test/OpenTelemetry.Tests/Shared/SkipOnWindowsRunnerTheoryAttribute.cs
Comment (alanwest) [reply test/OpenTelemetry.Tests/Shared/SkipOnWindowsRunnerTheoryAttribute.cs]: You'll probably need fetch tags to your fork. This should allow minver to properly version assemblies.

Comment (HofmeisterAn) [reply test/OpenTelemetry.Tests/Shared/SkipOnWindowsRunnerTheoryAttribute.cs]: Ok, I think we need to move the tests to another class, because the remaining tests trigger xUnit.net's `IAsyncLifetime` again. That's why we see the issues on Windows.

Comment (alanwest) [reply test/OpenTelemetry.Tests/Shared/SkipOnWindowsRunnerTheoryAttribute.cs]: Gotcha, yea moving the this one test to something like `SqlClientIntegrationTests.cs` would be fine.

Alternatively, instead of using IAsyncLifetime, could you just do the following in this one test?

```csharp
public async Task SuccessfulCommandTest(...)
{
    await this.databaseContainer.StartAsync();
    ...
    await this.databaseContainer.DisposeAsync().AsTask();
}
```

**Comment by alanwest** [comment]
Woot! @HofmeisterAn, I pulled down your branch and ran the tests on my M1 Mac within VSCode.

**Before starting up docker - test is skipped**
```text
OpenTelemetry.Instrumentation.SqlClient.Tests.SqlClientIntegrationTests.SuccessfulCommandTest:
    Outcome: Skipped
    Standard Output Messages:
    The Docker Linux engine is not available.
    
Total tests: 1. Passed: 0. Failed: 0. Skipped: 1
```

**After starting up docker - test runs and passes** (takes about 2 minutes first run. 1 minute subsequent runs.)
```text
[xUnit.net 00:00:01.03]   Starting:    OpenTelemetry.Instrumentation.SqlClient.Tests
[xUnit.net 00:01:57.99]   Finished:    OpenTelemetry.Instrumentation.SqlClient.Tests
----- Test Execution Summary -----

OpenTelemetry.Instrumentation.SqlClient.Tests.SqlClientIntegrationTests.SuccessfulCommandTest:
    Outcome: Passed
    
OpenTelemetry.Instrumentation.SqlClient.Tests.SqlClientIntegrationTests.SuccessfulCommandTest:
    Outcome: Passed
    
OpenTelemetry.Instrumentation.SqlClient.Tests.SqlClientIntegrationTests.SuccessfulCommandTest:
    Outcome: Passed
    
OpenTelemetry.Instrumentation.SqlClient.Tests.SqlClientIntegrationTests.SuccessfulCommandTest:
    Outcome: Passed
    
OpenTelemetry.Instrumentation.SqlClient.Tests.SqlClientIntegrationTests.SuccessfulCommandTest:
    Outcome: Passed
    
OpenTelemetry.Instrumentation.SqlClient.Tests.SqlClientIntegrationTests.SuccessfulCommandTest:
    Outcome: Passed
    
OpenTelemetry.Instrumentation.SqlClient.Tests.SqlClientIntegrationTests.SuccessfulCommandTest:
    Outcome: Passed
    
Total tests: 7. Passed: 7. Failed: 0. Skipped: 0
```

File: test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientIntegrationTests.cs
Comment (alanwest) [inline test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientIntegrationTests.cs:1]: Our dotnet-format CI is failing on this file. It's got a BOM that needs to be removed. File format should be plain UTF-8

Comment (reyang) [reply test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientIntegrationTests.cs:1]: bad 👮‍♂️

Comment (alanwest) [reply test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientIntegrationTests.cs:1]: Call the BOM squad!

**Comment by alanwest** [comment]
Ran on my Windows machine too using Visual Studio. I got the test to pass... kinda. But Visual Studio's test discoverability is confused. Maybe something about the custom theory? It does not seem to find all the test cases - should be 7, only finds 1.

Screenshot 2023-03-24 at 4 45 36 PM

Maybe just my machine, but curious if you also ran into anything funky here.. I assume you're also running on Windows.. cause you know, BOMs 😆

File: test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientIntegrationTests.cs
Comment (alanwest) [inline test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientIntegrationTests.cs:45]: Doesn't need to be in this PR, but this test extends our build/test CI by about 1 minute. This will only get longer as we port our other integration tests to use TestContainers.

We could add filter like `dotnet test --filter Traits\!~IntegrationTests` here to exclude them:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/094124e9ee61466a5bbf182d923658cde2c2dcf5/.github/workflows/ci.yml#L41-L42

Then we could make a separate integration test job that just ran all tests with trait equal to `*IntegrationTests`

**Review by alanwest — APPROVED** [review]
Other than the minor dotnet-format issue, LGTM! One more minor thing, I think we can remove the docker-compose and Dockerfile from this test project now, right?

I'm also curious why test discovery is not working.. but this could just be a "me" problem..

File: test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientIntegrationTests.cs
Comment (HofmeisterAn) [reply test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientIntegrationTests.cs]: I saved the file without BOM.

File: test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientIntegrationTests.cs
Comment (HofmeisterAn) [reply test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientIntegrationTests.cs:45]: Usually, putting tests in different assemblies and running them parallel can improve performance. In some cases, running tests against a singleton instance of a container or parallel against multiple instances can also increase performance, although this depends on the specific test. However, at the moment, I believe that the additional test time will be mostly due to pulling the image.

**Comment by HofmeisterAn** [comment]
> I'm also curious why test discovery is not working.. but this could just be a "me" problem..

Even though I use Windows most of the time, I do not use Visual Studio. While there is an issue with xUnit.net's [Theory](https://xunit.net/faq/theory-data-stability-in-vs) when used with Visual Studio, I do not believe this relates to this issue. Test discovery works fine in Rider (macOS and Windows).

**Pull request merged**
